cdap-app-fabric/src/test/java/co/cask/cdap/scheduler/NotificationSubscriberServiceTest.java

```diff
@@ -0,0 +1,207 @@
+      RetryStrategies.retryOnConflict(20, 100)
+    );
+  }
+
+  @AfterClass
+  public static void stop() throws Exception {
+    if (messagingService instanceof Service) {
+      ((Service) messagingService).stopAndWait();
+    }
+  }
+
+  @Test
+  public void testRunWorkflow() throws Exception {
+    NotificationSubscriberService subscriberService = injector.getInstance(NotificationSubscriberService.class);
+    try {
+      subscriberService.startAndWait();
+      for (int i = 0; i < 5; i++) {
+        testNewPartition();
+      }
+    } finally {
+      subscriberService.stopAndWait();
+    }
+  }
+
+  private void testNewPartition() throws Exception {
+    AppFabricTestHelper.deployApplicationWithManager(AppWithMultipleWorkflows.class, TEMP_FOLDER_SUPPLIER);
+
+    final ProgramId workflow1 = APP_ID.program(ProgramType.WORKFLOW, AppWithMultipleWorkflows.SomeWorkflow.NAME);
+    final ProgramId workflow2 = APP_ID.program(ProgramType.WORKFLOW, AppWithMultipleWorkflows.AnotherWorkflow.NAME);
+    Assert.assertEquals(0, store.getRuns(workflow1, ProgramRunStatus.ALL, 0, Long.MAX_VALUE, Integer.MAX_VALUE).size());
+    Assert.assertEquals(0, store.getRuns(workflow2, ProgramRunStatus.ALL, 0, Long.MAX_VALUE, Integer.MAX_VALUE).size());
+    publishNotification(dataEventTopic, workflow1);
+    publishNotification(dataEventTopic, workflow2);
+
+    try {
+      waitForCompleteRuns(1, workflow1);
+      waitForCompleteRuns(1, workflow2);
+    } finally {
+      LOG.info("workflow1 runRecords: {}",
+               store.getRuns(workflow1, ProgramRunStatus.ALL, 0, Long.MAX_VALUE, Integer.MAX_VALUE));
+      LOG.info("workflow2 runRecords: {}",
+               store.getRuns(workflow2, ProgramRunStatus.ALL, 0, Long.MAX_VALUE, Integer.MAX_VALUE));
+    }
+    store.removeApplication(APP_ID);
+    transactional.execute(new TxRunnable() {
+      @Override
+      public void run(DatasetContext context) throws Exception {
+        getScheduleDataset(context).deleteSchedules(APP_ID);
+      }
+    });
+  }
+
+  private void waitForCompleteRuns(int numRuns, final ProgramId program)
+    throws InterruptedException, ExecutionException, TimeoutException {
+
+    Tasks.waitFor(numRuns, new Callable<Integer>() {
+      @Override
+      public Integer call() throws Exception {
+        return store.getRuns(program, ProgramRunStatus.COMPLETED, 0, Long.MAX_VALUE, Integer.MAX_VALUE).size();
+      }
+    }, 5, TimeUnit.SECONDS);
+  }
+
+  private void publishNotification(TopicId topicId, final ProgramId programId)
+    throws TopicNotFoundException, IOException, TransactionFailureException {
+
+    final String name = topicId.getTopic() + programId.getProgram();
+    final DatasetId datasetId = programId.getNamespaceId().dataset(name);
+    Notification notification =
+      new Notification(Notification.Type.PARTITION, ImmutableMap.of("datasetId", datasetId.toString()));
+    transactional.execute(new TxRunnable() {
```

Comment:
You should be able to leverage the Schedule interface here that Andreas recently implemented, right?